### PR TITLE
Nvcr 8 create invoice list page

### PR DESCRIPTION
**Pull request by KateYatsiuk**

### CODE REVIEW

File: invoicer_project/invoices/templates/invoices/invoices_list.html [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        <div id="wrapper" class="toggled-2">
+            {% include "menu_wrapper.html" %}
+            <div id="page-content-wrapper">
+                <div class="container-fluid xyz" id="other_elements_invoices"></div>
+            </div>
+        </div>
+        <script src="{% static 'js/bundle.js' %}"></script>
+        <script type="module" src="{% static 'js/index.js' %}"></script>
+        <script type="module" src="{% static 'js/invoices_list_request.js' %}"></script>
+        <script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
```
[the remark this answers is not in the document]
Comment (Vitalik2705) [reply invoicer_project/invoices/templates/invoices/invoices_list.html]: Done

**Review by Vitalik2705 — APPROVED** [review]

**Review by KateSydor — APPROVED** [review]

**Review by NadiiaNanivska — APPROVED** [review]

**Review by OstapKH — APPROVED** [review]

File: invoicer_project/static/js/invoices_list_request.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
+const host = "http://127.0.0.1:8000";
+import {obtainUserInitials, obtainNewAccessToken, addCheckboxesListener, getUserData, addDeleteButtonListeners} from './request_utils.js';
+import {search} from './request_utils.js';
+document.getElementById("search_bar").addEventListener('keyup', ()=> { search('invoice_name', 'client_list_item')});
+
+function createInvoiceListContent(data) {
+    for (let i = 0; i < data.length; i++) {
+        let invoiceName = data[i]["name"];
```
Comment (maxchv) [inline invoicer_project/static/js/invoices_list_request.js:8]: Is it possible to use here [destructing assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)?
For example:
```js
const {id, name, price, date_of_payment} = data[i]
```
or
```js
const {id: invoiceId, name: invoiceName, price: invoicePirce, date_of_payment: invoiceDate} = data[i]
```

File: invoicer_project/static/js/request_utils.js [hunk 3]
```diff
@@ -187,4 +187,20 @@ export function addDeleteButtonListeners(selectorName, url) {
...
         });
     });
-}
\ No newline at end of file
+}
+
+export function search(elements, elementDivs) {
+    let input = document.getElementById('search_bar').value;
+    input = input.toLowerCase();
+    let x = document.getElementsByClassName(elements);
```
Comment (maxchv) [inline invoicer_project/static/js/request_utils.js:195]: let's use self describe names of variables. what is x? is it coordinate?

Comment (maxchv) [reply invoicer_project/static/js/request_utils.js:195]: replace let to const

**Review by maxchv — CHANGES_REQUESTED** [review]

**Review by dmitru4ok — APPROVED** [review]

File: invoicer_project/static/js/request_utils.js (hunk 3 above)
Comment (Vitalik2705) [reply invoicer_project/static/js/request_utils.js:195]: Done

File: invoicer_project/static/js/invoices_list_request.js (hunk 2 above)
Comment (Vitalik2705) [reply invoicer_project/static/js/invoices_list_request.js:8]: Done

File: invoicer_project/invoicer_project/settings.py [hunk 4]
```diff
@@ -40,9 +40,11 @@
     'rest_framework',
     'rest_framework_simplejwt',
     'rest_framework_simplejwt.token_blacklist',
+    'reportlab',
```
Comment (maxchv) [inline invoicer_project/invoicer_project/settings.py:43]: please, add this dependency into `requirements.txt` file

File: invoicer_project/invoices/views.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        for i, data in enumerate(items.data):
+            item_id = str(i + 1)
+            name = data['item_name']
+            price = str(data['price'])
+            amount = str(data['amount'])
+            unit = str(data['unit'])
+            sum_price = str(data["price"]*data["amount"])
+
+            if i == 0:
+                p.rect(table_row_x, table_row_y, id_width, 20)
```
Comment (maxchv) [inline invoicer_project/invoices/views.py:92]: what is 20? it is magic number. please, don't use magic number anymore. replace it to const

File: invoicer_project/invoices/views.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                p.rect(table_row_x + id_width, table_row_y, name_width, 20)
+                p.rect(table_row_x + id_width + name_width, table_row_y, amount_width, 20)
+                p.rect(table_row_x + id_width + name_width + amount_width, table_row_y, unit_width, 20)
+                p.rect(table_row_x + id_width + name_width + amount_width + unit_width, table_row_y, price_width, 20)
+                p.rect(table_row_x + id_width + name_width + amount_width + unit_width + price_width, table_row_y,
+                       sum_width, 20)
+
+                p.drawString(table_row_x + 5, table_row_y + 5, table_headers[0])
```
Comment (maxchv) [inline invoicer_project/invoices/views.py:100]: the same 5

File: invoicer_project/invoices/views.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+    def destroy(self, request):
+        user = get_user_from_jwt(request.headers)
+        invoice_ids = request.data['elementsIds']
+        invoices = Invoice.objects.filter(id__in=invoice_ids, user_id=user['user_id'])
+        invoices.delete()
+        return JsonResponse({}, status=204)
+
+    def download_data(self, request, invoice_id):
+        try:
```
Comment (maxchv) [inline invoicer_project/invoices/views.py:50]: I suggest to you move this code into separate module, for example, `invoice.py`.

File: invoicer_project/invoices/views.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+            p.drawString(table_row_x + id_width + name_width + amount_width + unit_width + 5, table_row_y + 5, price)
+            p.drawString(table_row_x + id_width + name_width + amount_width + unit_width + price_width + 5, table_row_y + 5, sum_price)
+            table_row_y -= 20
+
+        p.setFont('TimesNewRoman', 14)
+        p.drawString(375, table_row_y - 20,f"Всього: {serializer.data['price']} {serializer.data['currency']}")
+        p.line(40, table_row_y - 50, 565, table_row_y - 50)
+        p.drawString(40, table_row_y - 93, f"Від постачальника")
+        p.drawString(310, table_row_y - 93, f"Отримав(ла)")
+        p.line(40, table_row_y - 125, 280, table_row_y - 125)
```
Comment (maxchv) [inline invoicer_project/invoices/views.py:132]: 40? 125?

File: invoicer_project/invoicer_project/settings.py (hunk 4 above)
Comment (Vitalik2705) [reply invoicer_project/invoicer_project/settings.py:43]: Done

File: invoicer_project/invoices/views.py (hunk 5 above)
Comment (Vitalik2705) [reply invoicer_project/invoices/views.py:88]: Done

File: invoicer_project/invoices/views.py (hunk 6 above)
Comment (Vitalik2705) [reply invoicer_project/invoices/views.py:96]: Done

File: invoicer_project/invoices/views.py (hunk 8 above)
Comment (Vitalik2705) [reply invoicer_project/invoices/views.py:128]: Done

File: invoicer_project/invoices/views.py (hunk 7 above)
Comment (Vitalik2705) [reply invoicer_project/invoices/views.py:46]: Done

File: invoicer_project/invoices/invoice.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
+from io import BytesIO
+from reportlab.pdfgen import canvas
+from reportlab.pdfbase import pdfmetrics
+from reportlab.pdfbase.ttfonts import TTFont
+from reportlab.lib.pagesizes import letter
+def create_pdf(serializer, items):
```
Comment (maxchv) [inline invoicer_project/invoices/invoice.py:6]: please, follow the pep8 code conventions. Add spaces before function definition

Comment (maxchv) [reply invoicer_project/invoices/invoice.py:6]: I suggest you to use r[eformat code in pycharm ](https://www.jetbrains.com/help/pycharm/reformat-and-rearrange-code.html)

**Review by maxchv — CHANGES_REQUESTED** [review]

File: invoicer_project/invoices/invoice.py (hunk 9 above)
Comment (Vitalik2705) [reply invoicer_project/invoices/invoice.py:6]: Done

**Review by maxchv — APPROVED** [review]

**Pull request merged**